### ISSUE TITLE
feat: integrate State Space Models (SSMs) for event-based object detection

### ISSUE THREAD

**Issue by tallamjr** [tallamjr/evlib#29]
# feat: integrate State Space Models (SSMs) for event-based object detection

## Summary
Integrate Selective State Space Models (SSMs) into evlib following the CVPR 2024 paper "State Space Models for Event Cameras" by Zubic et al. This will provide a more efficient alternative to RNNs/LSTMs for processing sequential event data while maintaining comparable accuracy.

## Reference
- **Paper**: [State Space Models for Event Cameras](https://arxiv.org/abs/2402.15584) (CVPR 2024 Spotlight)
- **Reference Implementation**: Available in `libs/ssms_event_cameras/`
- **YouTube Video**: [CVPR 2024 Presentation](https://www.youtube.com/watch?v=WRZZJn6Me9M)

## Motivation
SSMs offer significant advantages over traditional recurrent models:
- **Parallelisable training**: No sequential bottleneck during training
- **Efficient inference**: Linear scaling with sequence length
- **Long-range dependencies**: Better at capturing temporal patterns in event streams
- **State-of-the-art performance**: Matches RVT accuracy with improved efficiency

### Core SSM Components
1. **S5 Layer** (`S5`): Core state space model with parallel scanning algorithm
2. **S5 Block** (`S5Block`): Transformer-style block with SSM replacing self-attention
3. **SSM-ViT Backbone**: Hierarchical Vision Transformer using S5 layers
4. **State Management**: Efficient handling of hidden states across sequences

### Integration Points
Following the established pattern in `python/evlib/models/`:
- **SSM Models**: `ssm.py`, `ssm_backbone.py`, `ssm_layers.py`
- **Detection Integration**: Reuse existing YOLOX head (`yolox_head.py`)
- **Configuration**: `SSMModelConfig` extending base `ModelConfig`
- **Weights**: Pre-trained checkpoints for Gen1/Gen4 datasets

### Phase 1: Core SSM Integration
- [ ] Implement `S5` core layer with parallel scanning
- [ ] Implement `S5Block` with transformer-style architecture
- [ ] Add JAX-compatible associative scan operations
- [ ] Implement SSM state initialisation functions

### Phase 2: SSM-ViT Architecture
- [ ] Implement `SSMBackbone` with 4-stage hierarchical structure
- [ ] Replace RNN/LSTM components with S5 blocks
- [ ] Add MaxViT integration for spatial-temporal processing
- [ ] Implement state carrying across sequence chunks

### Phase 3: Model Integration
- [ ] Create `SSM` main model class extending `BaseModel`
- [ ] Implement `SSMModelConfig` with variant support (base, small)
- [ ] Add preprocessing for event representations (voxel grids, stacked histograms)
- [ ] Integrate with existing YOLOX detection head

### Phase 4: Pre-trained Models & Evaluation
- [ ] Download and integrate pre-trained checkpoints
- [ ] Add model loading utilities and weight management
- [ ] Implement evaluation scripts for Gen1/Gen4 benchmarks
- [ ] Add performance comparison with RVT baselines

## Expected API
```python
import evlib

# Initialize SSM model
config = evlib.models.SSMModelConfig.base()
model = evlib.models.SSM(config=config, pretrained=True)

# Process events for detection
events = evlib.load_events("path/to/events.h5")
detections = model.detect(events, confidence_threshold=0.5)

# State-aware processing for streaming
state = model.init_state()
for chunk in event_stream:
    detections, state = model.detect_streaming(chunk, state)
```

### Reference to evlib Integration
```
libs/ssms_event_cameras/RVT/models/layers/s5/
→ python/evlib/models/ssm_layers.py

libs/ssms_event_cameras/RVT/models/detection/recurrent_backbone/maxvit_rnn.py
→ python/evlib/models/ssm_backbone.py

Integration with existing yolox_head.py, yolox_fpn.py
→ python/evlib/models/ssm.py (main model)
```

### Key Technical Components to Port
- **S5 Core**: Parallel scanning algorithm with associative scan
- **S5Block**: Transformer block with state space layers
- **MaxViT Integration**: Spatial-temporal processing with SSM layers
- **State Management**: Efficient hidden state handling for streaming

### Implementation Challenges & Solutions
1. **JAX Dependencies**: Port associative scan to pure PyTorch
2. **Complex Number Operations**: Handle complex-valued computations in S5
3. **State Management**: Efficient state passing for streaming scenarios
4. **Weight Compatibility**: Ensure pre-trained model loading works correctly

## Dependencies
- **Core**: PyTorch ≥2.0.0, einops, torch.compile support
- **Optional**: JAX functions for associative scanning (fallback to PyTorch)
- **Existing**: Leverage current YOLOX, MaxViT implementations

### 1 Mpx Dataset
- **S5-ViT-Base**: [gen4_base.ckpt](https://download.ifi.uzh.ch/rpg/CVPR24_Zubic/gen4_base.ckpt)
- **S5-ViT-Small**: [gen4_small.ckpt](https://download.ifi.uzh.ch/rpg/CVPR24_Zubic/gen4_small.ckpt)

### Gen1 Dataset
- **S5-ViT-Base**: [gen1_base.ckpt](https://download.ifi.uzh.ch/rpg/CVPR24_Zubic/gen1_base.ckpt)
- **S5-ViT-Small**: [gen1_small.ckpt](https://download.ifi.uzh.ch/rpg/CVPR24_Zubic/gen1_small.ckpt)

## Success Criteria
- [ ] SSM models load pre-trained weights successfully
- [ ] Detection accuracy matches reference implementation (±2% mAP)
- [ ] Training throughput improves over RVT baseline
- [ ] Integration follows evlib patterns (config, base classes, etc.)
- [ ] Comprehensive test coverage with real event data
- [ ] Documentation and examples for new SSM models

## Benefits Over RVT
- **Training Efficiency**: Parallel processing removes sequential bottleneck
- **Inference Speed**: Linear scaling vs quadratic for transformers
- **Memory Efficiency**: Constant memory usage for sequences
- **Long-range Modeling**: Better temporal dependency capture

## References
- [CVPR 2024 Paper](https://arxiv.org/abs/2402.15584)
- [Original JAX S5 Implementation](https://github.com/lindermanlab/S5)
- [Event Camera Datasets](https://dsec.ifi.uzh.ch/)
- [CVPR 2024 Poster](https://download.ifi.uzh.ch/rpg/CVPR24_Zubic/Zubic_CVPR24_poster.pdf)
